interfaces/builtin/custom_device.go

```diff
@@ -0,0 +1,417 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/udev"
+	"github.com/snapcore/snapd/interfaces/utils"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+)
+
+const customDeviceSummary = `provides access to custom devices specified via the gadget snap`
+
+const customDeviceBaseDeclarationSlots = `
+  custom-device:
+    allow-installation: false
+    allow-connection:
+      plug-attributes:
+        content: $SLOT(custom-device)
+    deny-auto-connection: true
+`
+
+var (
+	// A cryptic, uninformative error message that we use only on impossible code paths
+	customDeviceInternalError = errors.New(`custom-device interface internal error`)
+
+	// Validating regexp for filesystem paths
+	customDevicePathRegexp = regexp.MustCompile(`^/[^"@]*$`)
+
+	// Validating regexp for udev device names.
+	// We forbid:
+	// - `|`: it's valid for udev, but more work for us
+	// - `{}`: have a special meaning for AppArmor
+	// - `"`: it's just dangerous (both for udev and AppArmor)
+	// - `\`: also dangerous
+	customDeviceUDevDeviceRegexp = regexp.MustCompile(`^/dev/[^"|{}\\]+$`)
+
+	// Validating regexp for udev tag values (all but kernel devices)
+	customDeviceUDevValueRegexp = regexp.MustCompile(`^[^"{}\\]+$`)
+)
+
+// customDeviceInterface allows sharing customDevice between snaps
+type customDeviceInterface struct{}
+
+func (iface *customDeviceInterface) validateFilePath(path string, attrName string) error {
+	if !customDevicePathRegexp.MatchString(path) {
+		return fmt.Errorf(`custom-device %q path must start with / and cannot contain special characters: %q`, attrName, path)
+	}
+
+	if !cleanSubPath(path) {
+		return fmt.Errorf(`custom-device %q path is not clean: %q`, attrName, path)
```

Comment:
The error message is a bit cryptic for someone who isn't familiar with filepath.Clean(), but I see we have a similar problem in other places (mount-control, shared-mem). Perhaps this could be worded differently in a followup? eg:
```
custom-device %q path must not contain repeated / or .. 
```

Reply:
I will do this in a follow-up, for all the interfaces that use a similar pattern. Though I didn't find it cryptic at all when I first saw it, despite not being familiar with `filepath.Clean()` myself (it might be because the QtCore library has a similar method :-) ).

It's also going to be a bit hard to find out an error message that is clear enough: note that a sequence of `../` is allowed at the beginning of the path, but not in the middle of it. And if I'm not mistaken `./` is not allowed anywhere. So, it may be that leaving the generic text "not clean" is easier for the developer (who can anyway do a web search for it) than giving out a precise error message which though risks being more convoluted.

Reply:
Python has os.path.normpath() (aka normalized path), I think it can also be referred to as canonical path? Anyway, maybe it's enough to indicate that the path cannot contain `.`, `..` or repeated `/`.